### DISCUSSION TITLE
Improved error handling?

### DISCUSSION THREAD

**Discussion by vmiklos** [luckasRanarison/kewb#9] [Ideas]
I think this is a decent summary on idiomatic error handling in Rust:

https://www.lurklurk.org/effective-rust/errors.html

TL;DR: users of the kewb library probably expect an error type that implements Error, while it seems you currently use String in many cases.

Some possible ways to improve on this:

1) If you're interested in fine-grained error codes, then the `thiserror` crate can help with that.

2) If you would like to do a minimal improvement, then using `anyhow` would be also better, even if that is primary intended for non-library code.

Do you have an opinion of any of these would be interesting?

My pet two functions to improve would be `kewb::FaceCube::try_from()` and `kewb::State::try_from()`, where I currently just use unwrap()s in https://github.com/vmiklos/vmexam/blob/16284e89c8cb3b95fd795b756af635375c6f3751/share-vmiklos-hu-apps/src/rubik.rs#L17-L19 -- because manual match on Ok/Err is boring, but easy error chaining is not working due to String not implementing Error. Thanks.

**Comment by luckasRanarison** [comment]
This was already done actually (7fd0a1e), but I just went with fmt::Display and Debug, I wasn't aware of the Error trait.

I haven't just updated the crate yet, I wanted to make some additional changes before doing it. It'll introduce some breaking changes. 

I was also thinking about changing the pruning and move data to another directory, currently I'm just using the current executable directory and it's not correct. Or should I allow the user the do it and add some parameters to `read_table` and `write_table`? And remove the automatic write if not found.

**Comment by vmiklos** [comment]
> but I just went with fmt::Display and Debug, I wasn't aware of the Error trait.

Oh, great. When I looked last time, most errors were String, that's why I started this discussion. :-)

Up to you, I found the amount of required boilerplate annoying and went with `thiserror`, but no problem with just providing the Error implementation yourself.

> Or should I allow the user the do it and add some parameters to read_table and write_table?

I think the ideal would be to keep these paths as parameters in the library crate. Then binary crate can either keep the current default or the more idiomatic way is to store this in the user profile in some way. E.g. Linux has the XDG spec for this ($XDG_CACHE_HOME, defaults to  $HOME/.cache, so  $HOME/.cache/kewb/ in your case), Windows also has the AppData directory for something similar.

**Reply by luckasRanarison** [reply]
> Up to you, I found the amount of required boilerplate annoying and went with `thiserror`, but no problem with just providing the Error implementation yourself.
> 
I know about thiserror but I didn't want to add more dependencies.

> I think the ideal would be to keep these paths as parameters in the library crate. Then binary crate can either keep the current default or the more idiomatic way is to store this in the user profile in some way. E.g. Linux has the XDG spec for this ($XDG_CACHE_HOME, defaults to  $HOME/.cache, so  $HOME/.cache/kewb/ in your case), Windows also has the AppData directory for something similar.

Thanks for the tips :)

**Reply by luckasRanarison** [reply]
Hi! Sorry for the long absence. Just in case you still want to change the code in your project because of the poor error handling of the last version, I've published a [new version](https://crates.io/crates/kewb) with better APIs and extra features (may not be relevant to your project though). You can also bundle the move and pruning table in the executable now, so now more read operations are needed.

**Reply by vmiklos** [reply]
Thanks, I just updated, so far so good.

Regarding ergonomic errors, one pain point still is that `std::error::Error` is not implemented for `kewb::error::Error`.

You're almost there, since you implement `fmt::Display`, so you have a way to turn errors into strings. What's missing is something like this:

```rust
impl std::error::Error for Error {
    fn description(&self) -> &str {
        ...
    }
}
```

See https://docs.rs/thiserror/latest/thiserror/ if you want to do this with a macro; or if you don't mind a bit of redundant code, it's possible to also manually do this. Or if you have an opinion which way you prefer, I can also look at doing a PR to add this. Thanks. :-)

**Comment by luckasRanarison** [comment]
Thanks for the remarks, implementing the `Error` trait manually is a lot of work so I went with `thiserror`.
I've already made a [patched](https://github.com/luckasRanarison/kewb/releases/tag/v0.4.1) version :)

**Reply by vmiklos** [reply]
Cool, this does what I had in mind! Thanks.
